### PR TITLE
Update S3.js

### PR DESCRIPTION
**Pull request by alexkruegger**
There is a bug in the S3 StorageProvider implementation. According to Amazon API you have to check res.StatusCode to be sure that request has been successfully processed. In the current Master branch version of the S3 provider, the 'save' / 'remove' function return no error even if you provide incorrect API keys.

BugFix is to check StatusCode and throw an Error with res.body that contain stringified object with Error description in case of StatusCode not in [200-209].

### CODE REVIEW

File: lib/S3.js [hunk 1]
```diff
@@ -30,22 +30,43 @@ var S3StorageProvider = function(options) {
     bucket: this._options.bucket,
     region: this._options.region
   })
-}
+
+S3StorageProvider.prototype._cb = function(error, res, callback) {
+  var body = '';
+
+  res.on('data', function (chunk) {
+    body += chunk;
```
Comment (achingbrain) [inline lib/S3.js]: How big are these responses likely to be? The whole thing is stored in memory here, just so we can pass it into the error which might be problematic at large volumes. Would it be sufficient to just act on `res.statusCode` instead of streaming the body too?  Or maybe only streaming the body if the code wasn't 2xx?

**Comment by alexkruegger** [comment]
Well, The body is important as it contains Error description like "Invalid API keys" and so on. Without body we only have StatusCode = 400 and don't know what exactly happened.
The body itself is small, I implement "chunk"-ed style of body content accumulation according to my habbit ). So streaming the body is important only if we have an error (status code != 2xx).

**Comment by achingbrain** [comment]
My point is that the body gets streamed even if we're not going to use it.  I think it would be better to do something like:

```javascript
S3StorageProvider.prototype._cb = function(error, res, callback) {
  if (error) {
    return callback(error);
  }

  if (res.statusCode >= 200 && res.statusCode <= 209) {
    return callback(null, res.req.url);
  }

  var body = '';

  res.on('data', function (chunk) {
    body += chunk;
  });

  res.on('end', function () {
    var err = new Error();
    err.statusCode = res.statusCode;
    err.body = body;

    callback(err);
  });
};
```

**Comment by alexkruegger** [comment]
Yes, I see. Agree, you code looks better ) Can I ask you to apply the patch?

**Comment by achingbrain** [comment]
If you could fix the failing tests without lowering the coverage I'll merge it.

**Comment by alexkruegger** [comment]
Alex, I have no idea why my patch can't pass the ['  - should store and remove a file'] test. Can I ask you to assist me?
Two words about patch - I have rebuild the code of _cb function to meet the test's condition, but now it looks very complexity. Possible the best way is to change the S3Test.js to reflect the fact that callback of .PutFile / .DeleteFile has a 'res' argument. ???

Sorry for possible stupid question, it is my first Pull request here, on the GH.

**Comment by achingbrain** [comment]
Congratulations on the first PR!

You are nearly there - Travis has passed so all existing tests pass, however Coveralls [reports](https://coveralls.io/builds/2660915/source?filename=lib%2FS3.js) that code coverage has decreased from 100% to 74.42%, meaning that you've introduced new code that is not covered by unit tests.

You need to write some extra tests to cover the code that's being missed.

You can check the code coverage before committing by running `npm test` in the root of this repo on your local machine.

**Comment by alexkruegger** [comment]
Hi!
I have written tests to cover 100% of new code. Please check that all is OK.
Alex.

File: .npmignore [hunk 2]
```diff
@@ -0,0 +1,7 @@
+*.iml
+.idea
+coverage
+.DS_Store
+Thumbs.db
+node_modules
+*.log
```
Comment (achingbrain) [inline .npmignore]: Is this file necessary? The npm docs [say](https://docs.npmjs.com/misc/developers#keeping-files-out-of-your-package) that if there's no `.npmignore` it'll use the `.gitignore` file instead.

File: lib/S3.js [hunk 3]
```diff
@@ -30,22 +30,62 @@ var S3StorageProvider = function(options) {
     bucket: this._options.bucket,
     region: this._options.region
   })
-}
+
+this._cb = function(error, res, callback) {
```
Comment (achingbrain) [inline lib/S3.js]: Could this field please have a more descriptive name?  `this._cb` doesn't tell me much about what it's supposed to do.

File: lib/S3.js [hunk 4]
```diff
@@ -30,22 +30,62 @@ var S3StorageProvider = function(options) {
...
+
+  if (error) {
+    return callback(error);
```
Comment (achingbrain) [inline lib/S3.js]: Please, to keep the code consistent, no semi-colons.

File: lib/S3.js [hunk 5]
```diff
@@ -30,22 +30,62 @@ var S3StorageProvider = function(options) {
...
 S3StorageProvider.prototype.remove = function(attachment, callback) {
+  var self = this;
+
   if(!attachment.url) {
-    return callback()
+    return callback();
   }
 
-  this._client.deleteFile(path.basename(attachment.url), callback)
+//  this._client.deleteFile(path.basename(attachment.url), callback)
```
Comment (achingbrain) [inline lib/S3.js]: No commented out code, please. Version control takes care of that.

File: package.json [hunk 6]
```diff
@@ -1,5 +1,7 @@
 {
-  "author": "Alex Potsides",
+  "author": {
+    "name": "Alex Potsides"
+  },
```
Comment (achingbrain) [inline package.json]: Why is this change necessary?

File: package.json [hunk 7]
```diff
@@ -49,5 +51,30 @@
...
+      "name": "achingbrain",
+      "email": "alex@achingbrain.net"
+    }
+  ],
+  "dist": {
+    "shasum": "79039939b81c7453e8d9f7a4b8c098c7bcc5fc71",
+    "tarball": "http://registry.npmjs.org/mongoose-crate-s3/-/mongoose-crate-s3-1.1.1.tgz"
+  },
+  "directories": {},
+  "_resolved": "https://registry.npmjs.org/mongoose-crate-s3/-/mongoose-crate-s3-1.1.1.tgz"
```
Comment (achingbrain) [inline package.json]: There's a lot of noise here - looks like you've taken the generated package.json file from an npm download.

File: test/S3Test.js [hunk 8]
```diff
@@ -1,285 +1,343 @@
...
-  fs = require('fs'),
-  path = require('path'),
-  request = require('request'),
-  async = require('async'),
-  S3 = require('../lib/S3'),
-  sinon = require('sinon'),
-  proxyquire = require('proxyquire')
-
-var bucket =  process.env.CRATE_BUCKET
+        crypto = require('crypto'),
```
Comment (achingbrain) [inline test/S3Test.js]: Please don't reformat the code.

File: test/S3Test.js [hunk 9]
```diff
@@ -1,285 +1,343 @@
...
-  })
-
-  it('should set a default acl', function(done) {
-    var s3 = new S3({
-      key: 'foo',
-      secret: 'bar',
-      bucket: 'baz'
+describe('S3', function () {
+
+    it('should require options', function (done) {
```
Comment (achingbrain) [inline test/S3Test.js]: Again, all the two-space indentation seems to have changed to four-spaces.  I'll probably run [standard](https://github.com/feross/standard) in a pre-commit hook in future to catch this stuff.

File: test/S3Test.js [hunk 10]
```diff
@@ -1,285 +1,343 @@
...
-  it('should not remove a file with no URL', function(done) {
-    var client = {
-      deleteFile: sinon.stub()
-    }
 
-    var S3 = proxyquire('../lib/S3', {
-      'knox': {
-        createClient: function () {
-          return client
+    // Added by Alex Kruegger
```
Comment (achingbrain) [inline test/S3Test.js]: This is unnecessary - the git log holds this information.

File: test/S3Test.js [hunk 11]
```diff
@@ -1,285 +1,343 @@
...
+    it('should pass the tests for fixed bug', function (done) {
```
Comment (achingbrain) [inline test/S3Test.js]: Can you please make the test name more descriptive?

File: test/S3Test.js [hunk 12]
```diff
@@ -1,285 +1,343 @@
...
 
-    var s3 = new S3({
-      key: 'PUT_YOUR_KEY_HERE',
-      secret: 'PUT_YOUR_BUCKET_HERE',
-      bucket: 'PUT_YOUR_BUCKET_HERE',
-      region: 'PUT_YOUR_REGION_HERE'
+        var S3 = proxyquire('../lib/S3', {
+            'knox': {
+                createClient: function () {
+                    return client
```
Comment (achingbrain) [inline test/S3Test.js]: `client` has a stubbed `_cb` field - is this necessary?  Usually you'd do this because you want to stub some behaviour but you don't seem to do anything with `client._cb` after this.

**Comment by alexkruegger** [comment]
Sorry for dirty code, I have fixed it. And did some refactoring according to your comments.

**Comment by achingbrain** [comment]
Much better! It's failing because travis is having problems connecting to coveralls - don't worry about it, as long as the tests pass and are at 100% coverage that's fine.

Could you please squash all the commits into one and I'll merge this and publish the update.

**Comment by alexkruegger** [comment]
I'm not so nice in git at this moment so please advice me - to squash all commits into one from root and push them I need to run:
git rebase --interactive --root
git push
?

**Comment by achingbrain** [comment]
You want to run an interactive rebase like `git rebase -i HEAD~N` where N is the number of commits to rewrite.  This PR has 13 commits so 13 is the number to use:

```sh
$ git rebase -i HEAD~13
```

You'll then see something like this:

```
pick c6d6747 Update S3.js
pick a4248d9 Update S3.js
pick 0fc93db working...
pick 1e677f4 working...
pick 0566bb6 working...
pick 190289b working...
pick cfcc0ba Update S3.js
pick b6dec3b Update S3.js
pick 57c187b working commit
pick 0654846 working commit
pick 7efc5fa working commit
pick a791f6d fixed callback function name
pick 822a094 refactor code to meet requirements

# Rebase 88a104c..822a094 onto 88a104c
#
# Commands:
#  p, pick = use commit
#  r, reword = use commit, but edit the commit message
#  e, edit = use commit, but stop for amending
#  s, squash = use commit, but meld into previous commit
#  f, fixup = like "squash", but discard this commit's log message
#  x, exec = run command (the rest of the line) using shell
...
```

You want to use the `s, squash = use commit, but meld into previous commit` command.

Make the file look like this (`s` is shorthand for `squash`):

```
pick c6d6747 Update S3.js
s a4248d9 Update S3.js
s 0fc93db working...
s 1e677f4 working...
s 0566bb6 working...
s 190289b working...
s cfcc0ba Update S3.js
s b6dec3b Update S3.js
s 57c187b working commit
s 0654846 working commit
s 7efc5fa working commit
s a791f6d fixed callback function name
s 822a094 refactor code to meet requirements
```

Save the file and exit and the rebase will be performed.  You'll then be prompted to update the commit message for the newly squashed commit - do that, save the file and exit.

After that do a `git push --force` and it'll replace the 13 commits with the one from your local history.

**Comment by alexkruegger** [comment]
Thank you for detailed instruction. I mean I did it. :)

**Comment by achingbrain** [comment]
I think you may have merged a commit too far - see how it's included the version number rev from the current HEAD of master?

**Comment by alexkruegger** [comment]
Yes my mistake, sorry. Is it correct now?

**Comment by achingbrain** [comment]
Yep, that's great - thanks for all your work.

**Comment by alexkruegger** [comment]
Thank you for nice module and for support!

**Comment by achingbrain** [comment]
Published as v1.1.2
